llvm/lib/Transforms/Scalar/InferAlignment.cpp

```diff
@@ -35,8 +37,39 @@ static bool tryToImproveAlign(
       return true;
     }
   }
-  // TODO: Also handle memory intrinsics.
-  return false;
+
+  IntrinsicInst *II = dyn_cast<IntrinsicInst>(I);
+  if (!II)
+    return false;
+
+  // TODO: Handle more memory intrinsics.
+  switch (II->getIntrinsicID()) {
+  case Intrinsic::masked_load:
+  case Intrinsic::masked_store: {
+    Value *PtrOp = II->getIntrinsicID() == Intrinsic::masked_load
+                       ? II->getArgOperand(0)
+                       : II->getArgOperand(1);
+    Value *AlignOp = II->getIntrinsicID() == Intrinsic::masked_load
+                         ? II->getArgOperand(1)
+                         : II->getArgOperand(2);
+
+    Align OldAlign = cast<ConstantInt>(AlignOp)->getAlignValue();
+    Align PrefAlign = getKnownAlignment(PtrOp, DL, II);
```

Comment:
This should be `DL.getPrefTypeAlign()` of the value type (return type for load, op0 for store).